### PR TITLE
optimizer: use multi-column index

### PR DESCRIPTION
**Pull request by coocood**
fixes https://github.com/pingcap/tidb/issues/859

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @qiuyesuifeng PTAL

File: optimizer/plan/refiner.go [hunk 1]
```diff
@@ -56,12 +56,25 @@ var fullRange = []rangePoint{
...
+		p.Ranges = rb.buildIndexRanges(point)
+		for i := 1; i < p.AccessEqualCount; i++ {
+			point = rb.build(p.AccessConditions[i])
+			p.Ranges = rb.appendIndexRanges(p.Ranges, point)
+		}
+	}
 	rangePoints := fullRange
-	for _, cond := range p.AccessConditions {
-		rangePoints = rb.intersection(rangePoints, rb.build(cond))
+	// Build rangePoints for non-equal access condtions.
```
Comment (qiuyesuifeng) [inline optimizer/plan/refiner.go]: conditions

File: optimizer/plan/plan_test.go [hunk 2]
```diff
@@ -585,3 +585,32 @@ func (s *testPlanSuite) TestJoinPath(c *C) {
...
+		mockResolve(stmt)
+		b := &planBuilder{}
+		p := b.buildFrom(stmt)
+		err = Refine(p)
+		c.Assert(err, IsNil)
+		idxScan, ok := p.(*IndexScan)
+		c.Assert(ok, IsTrue)
+		c.Assert(idxScan.AccessEqualCount, Equals, ca.accessEqualCount)
+		c.Log(idxScan.Ranges[0].LowVal)
+		c.Assert(len(idxScan.Ranges[0].LowVal), Equals, ca.usedColumnCount)
```
Comment (qiuyesuifeng) [inline optimizer/plan/plan_test.go]: Using HasLen?

**Comment by shenli** [comment]
LGTM

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**